From 06324c31ac6db4f9202a098bca74c87e985dd862 Mon Sep 17 00:00:00 2001
From: Yonghua Zheng <yonghua.zheng@esimtek.com>
Date: Thu, 6 Jul 2017 17:49:39 +0800
Subject: [PATCH] ion: add ion_getphyaddr for sun7i

Sun7i specific ion function to reterieve physical address
for a given memory. Required and used by gralloc.sun7i
hardware module.

Change-Id: I844f87bfda549c6ca840dd0108e01c06982bb200
---
 include/ion/ion.h |  2 ++
 libion/ion.c      | 17 +++++++++++++++++
 2 files changed, 19 insertions(+)

diff --git a/include/ion/ion.h b/include/ion/ion.h
index 018c0a1..2889a9c 100644
--- a/include/ion/ion.h
+++ b/include/ion/ion.h
@@ -37,6 +37,8 @@ int ion_map(int fd, struct ion_handle *handle, size_t length, int prot,
             int flags, off_t offset, unsigned char **ptr, int *map_fd);
 int ion_share(int fd, struct ion_handle *handle, int *share_fd);
 int ion_import(int fd, int share_fd, struct ion_handle **handle);
+// sunxi
+int ion_getphyaddr(int fd, void *handle);
 
 __END_DECLS
 
diff --git a/libion/ion.c b/libion/ion.c
index 020c35b..0c787cc 100644
--- a/libion/ion.c
+++ b/libion/ion.c
@@ -154,3 +154,20 @@ int ion_sync_fd(int fd, int handle_fd)
     };
     return ion_ioctl(fd, ION_IOC_SYNC, &data);
 }
+
+int ion_getphyadr(int fd, void *handle)
+{
+    struct ion_custom_data custom_data;
+    struct sunxi_phys_data {
+        void *handle;
+        unsigned int phys_addr;
+        unsigned int size;
+    } phys_data;
+    custom_data.cmd = 7 ; // ION_IOC_SUNXI_PHYS_ADDR;
+    custom_data.arg = (unsigned long)&phys_data;
+    phys_data.handle = handle;
+    int ret = ioctl(fd, ION_IOC_CUSTOM, &custom_data);
+    if (ret < 0)
+        return 0;
+    return phys_data.phys_addr;
+}
-- 
2.7.4

